cpp/src/arrow/sparse_tensor.h

```diff
@@ -136,50 +139,80 @@ class ARROW_EXPORT SparseCOOIndex : public internal::SparseIndexBase<SparseCOOIn
   std::shared_ptr<Tensor> coords_;
 };
 
-// ----------------------------------------------------------------------
-// SparseCSRIndex class
+namespace internal {
 
-/// \brief EXPERIMENTAL: The index data for a CSR sparse matrix
-///
-/// A CSR sparse index manages the location of its non-zero values by two
-/// vectors.
-///
-/// The first vector, called indptr, represents the range of the rows; the i-th
-/// row spans from indptr[i] to indptr[i+1] in the corresponding value vector.
-/// So the length of an indptr vector is the number of rows + 1.
-///
-/// The other vector, called indices, represents the column indices of the
-/// corresponding non-zero values.  So the length of an indices vector is same
-/// as the number of non-zero-values.
-class ARROW_EXPORT SparseCSRIndex : public internal::SparseIndexBase<SparseCSRIndex> {
- public:
-  static constexpr SparseTensorFormat::type format_id = SparseTensorFormat::CSR;
+struct SparseMatrixCompressedAxis {
+  /// EXPERIMENTAL: The axis to be compressed
+  enum type {
+    /// The value for CSR matrix
+    ROW,
+    /// The value for CSC matrix
+    COLUMN
+  };
+};
+
+ARROW_EXPORT
+Status ValidateSparseCSXIndex(const std::shared_ptr<DataType>& indptr_type,
+                              const std::shared_ptr<DataType>& indices_type,
+                              const std::vector<int64_t>& indptr_shape,
+                              const std::vector<int64_t>& indices_shape,
+                              char const* type_name);
 
-  /// \brief Make SparseCSRIndex from raw properties
-  static Result<std::shared_ptr<SparseCSRIndex>> Make(
+ARROW_EXPORT
+void CheckSparseCSXIndexValidity(const std::shared_ptr<DataType>& indptr_type,
+                                 const std::shared_ptr<DataType>& indices_type,
+                                 const std::vector<int64_t>& indptr_shape,
+                                 const std::vector<int64_t>& indices_shape,
+                                 char const* type_name);
+
+template <typename SparseIndexType, SparseMatrixCompressedAxis::type COMPRESSED_AXIS>
+class SparseCSXIndex : public SparseIndexBase<SparseIndexType> {
+ public:
+  static constexpr SparseTensorFormat::type format_id = std::conditional<
+      COMPRESSED_AXIS == SparseMatrixCompressedAxis::ROW,
+      std::integral_constant<SparseTensorFormat::type, SparseTensorFormat::CSR>,
```

Comment:
Why is `std::integral_constant` needed?

Reply:
Oh, it's enough to use a ternary operator!